python/pyarrow/array.pxi

```diff
@@ -84,6 +84,18 @@ cdef _ndarray_to_array(object values, object mask, DataType type,
         return pyarrow_wrap_array(chunked_out.get().chunk(0))
 
 
+def _handle_arrow_array_protocol(obj, type, mask, size):
+    if mask is not None or size is not None:
+        raise ValueError(
+            "Cannot specify a mask or a size when passing an object that is "
+            "converted with the __arrow_array__ protocol.")
+    res = obj.__arrow_array__(type=type)
+    if not isinstance(res, Array):
+        raise ValueError("The object's __arrow_array__ method does not "
```

Comment:
`TypeError`, no?

Reply:
Numpy returns a ValueError in this case, but yes, a TypeError seems more appropriate